### PR TITLE
Add support for portable application layout

### PR DESCRIPTION
**Pull request by anurse**
A more betterer version of #1784 

Instead of `--portable`, we use `netstandard1.x` to identify portable apps (for now).

Haven't gotten a 100% clean build yet, but wanted to start the PR up.

/cc @pakrym @davidfowl

### CODE REVIEW

**Comment by weshaggard** [comment]
What is the spec'ed way we plan to represent this? (Sorry too many spec locations I do not know where the one covering this is at).

I ask because the TFM for these applications will end up being netcoreapp1.0 and not netstandard1.x.

**Comment by anurse** [comment]
@weshaggard The restriction on `netstandard` TFM is purely a point-in-time staging issue. NuGet does not infer RIDs for that TFM. Also, we haven't fully landed on the TFM name at all.

**Comment by anurse** [comment]
ping @davidfowl - since @pakrym is OOF and I'd like to get this in, I'd appreciate a brief review :). The builds should be green (enough) now.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectDependenciesCommandResolver.cs [hunk 1]
```diff
@@ -25,12 +25,12 @@ public class ProjectDependenciesCommandResolver : ICommandResolver
         {
             if (environment == null)
             {
-                throw new ArgumentNullException("environment");
+                throw new ArgumentNullException(nameof(environment));
```
Comment (anurse) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/ProjectDependenciesCommandResolver.cs]: VS has started complaining about these as errors now :). I kinda like that, but it means I feel obligated to update them :P

**Comment by anurse** [comment]
I'm going to merge this, but there is one more issue to resolve before we're completely ready to integrate, it's a small thing and I've mostly got a fix though. I'll file a bug for it.

**Comment by anurse** [comment]
https://github.com/dotnet/cli/issues/1828

File: TestAssets/TestProjects/TestAppWithWrapperProjectDependency/TestLibrary/README.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# What is this?
```
Comment (pakrym) [inline TestAssets/TestProjects/TestAppWithWrapperProjectDependency/TestLibrary/README.md]: :+1:

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs [hunk 3]
```diff
@@ -14,11 +14,16 @@
...
 using NuGet.Frameworks;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
 
 namespace Microsoft.Dotnet.Cli.Compiler.Common
 {
     public class Executable
     {
+        // GROOOOOSS
+        private static readonly string RedistPackageName = "Microsoft.NETCore.App";
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: Omg file a bug to remove?

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs [hunk 4]
```diff
@@ -115,6 +122,37 @@ private void WriteDepsFileAndCopyProjectDependencies(LibraryExporter exporter)
             CopyAssets(packageExports);
         }
 
+        private void WriteRuntimeConfig(LibraryExporter exporter)
+        {
+            if (!_context.TargetFramework.IsDesktop())
+            {
+                // TODO: Suppress this file if there's nothing to write? RuntimeOutputFiles would have to be updated
+                // in order to prevent breaking incremental compilation...
```
Comment (davidfowl) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: File bug

File: src/Microsoft.DotNet.ProjectModel/OutputPathsCalculator.cs [hunk 5]
```diff
@@ -70,11 +70,7 @@ public class OutputPathsCalculator
 
             var compilationFiles = new CompilationOutputFiles(compilationOutputPath, project, configuration, framework);
 
-            RuntimeOutputFiles runtimeFiles = null;
-            if (runtimeOutputPath != null)
-            {
-                runtimeFiles = new RuntimeOutputFiles(runtimeOutputPath, project, configuration, framework);
-            }
+            RuntimeOutputFiles runtimeFiles = new RuntimeOutputFiles(runtimeOutputPath, project, configuration, framework, runtimeIdentifier);
             return new OutputPaths(intermediateOutputPath, compilationOutputPath, runtimeOutputPath, compilationFiles, runtimeFiles);
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/OutputPathsCalculator.cs]: Var

File: src/Microsoft.DotNet.ProjectModel/ProjectContext.cs [hunk 6]
```diff
@@ -140,6 +140,12 @@ public static IEnumerable<ProjectContext> CreateContextForEachTarget(string proj
 
         public ProjectContext CreateRuntimeContext(IEnumerable<string> runtimeIdentifiers)
         {
+            // Temporary until we have removed RID inference from NuGet
+            if(TargetFramework.IsCompileOnly)
+            {
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/ProjectContext.cs]: Formatting

File: src/dotnet/commands/dotnet-build/CompileContext.cs [hunk 7]
```diff
@@ -446,24 +441,22 @@ private void MakeRunnable()
...
-                        File.Exists(Path.Combine(coreclrDir, "mscorlib.ni.dll")))
-                    {
-                        continue;
-                    }
-                    var mscorlibFile = exp.RuntimeAssemblies.FirstOrDefault(r => r.Name.Equals("mscorlib") || r.Name.Equals("mscorlib.ni")).ResolvedPath;
-                    File.Copy(mscorlibFile, Path.Combine(coreclrDir, Path.GetFileName(mscorlibFile)), overwrite: true);
+                    continue;
                 }
+                var mscorlibFile = exp.RuntimeAssemblies.FirstOrDefault(r => r.Name.Equals("mscorlib") || r.Name.Equals("mscorlib.ni")).ResolvedPath;
+                File.Copy(mscorlibFile, Path.Combine(coreclrDir, Path.GetFileName(mscorlibFile)), overwrite: true);
```
Comment (davidfowl) [inline src/dotnet/commands/dotnet-build/CompileContext.cs]: What is this hack? And where is the comment

File: src/Microsoft.DotNet.Compiler.Common/Executable.cs (hunk 3 above)
Comment (weshaggard) [inline src/Microsoft.DotNet.Compiler.Common/Executable.cs]: Yes please.

**Pull request merged**